### DISCUSSION TITLE
Modify send content

### DISCUSSION THREAD

**Discussion by mangfu26** [encode/httpx#2039] [Q&A] [answered]
I have an HTTPX client.
I want to encrypt the data sent inside Event Hooks, so I modified `request._content` directly, but the data sent doesn't seem to change, does anyone know why this is?

```python
# client 

import httpx


client = httpx.Client()


def __req(request: httpx.Request):
    request._content = b'Hello'


client.event_hooks['request'].append(__req)

response = client.post(
    'http://192.168.8.184:10240/test',
    data=b'Test'
)

client.close()
```

```python
# server

from fastapi import FastAPI, Request


app = FastAPI()

@app.post('/test')
async def test(request: Request):
    '''
    '''
    print(await request.body())

    return {'message': 'Hello'}
```

```bash
# output

INFO: Application startup complete.
b'Test'
INFO: 192.168.8.184:34930 - "POST /test HTTP/1.1" 200 OK

```

Translated with www.DeepL.com/Translator (free version)

**Comment by lovelydinosaur** [comment, marked as answer]
Working with private implementation details like that isn't going to be a good route to go down - you shouldn't expect it to work, and even if it did, you might well find that some future version broke it.

It's not clear what *actual* behaviour you're trying to implement, and why. We'd need a clearer idea of that in order to help you find a better approach.

**Reply by mangfu26** [reply]
I am trying to develop a microservice framework using FastAPI and HTTPX to encrypt data using RSA algorithm before sending HTTP requests, I am planning to implement RSA encryption in Event Hooks, if I can't implement it in Event Hooks, is there any other way, thanks.

**Reply by lovelydinosaur** [reply]
I'd probably aim to encrypt it *before* passing it to `httpx` (Eg. so it can set the correct `Content-Length` in cases where it knows that up front.)

Tho you could also look at [implementing a custom transport](https://www.python-httpx.org/advanced/#custom-transports) to modify the request before passing it to the connection pool.

**Reply by mangfu26** [reply]
Thank you for your patient answer, my problem was solved

**Reply by lovelydinosaur** [reply]
Glad to hear it - if you get a chance please do let us know the approach you took - might come in useful for someone else.
